backend/tests/data_layer/database/crud/test_crud_utils.py

```
@@ -0,0 +1,909 @@
+#################### TESTS ####################
+
+
+def validate_pre_upsert_data(
+    upsert_data: List[Dict[str, Any]], model: Type[SQLModel], session: Session
+) -> None:
+    """
+    Validate data before upsert operation.
+    """
+    for data in upsert_data:
+        prev_data = session.exec(
+            select(model).where(getattr(model, "token") == data["token"])
```

Ruff (0.11.9)

Do not call `getattr` with a constant attribute value. It is not any safer than normal property access.

Replace `getattr` with attribute access

(B009)